src/cdr/resolver.go

```diff
@@ -0,0 +1,271 @@
+package cdr
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/cipher"
+	"hash"
+	"io"
+	"net/http"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/chacha20"
+	"google.golang.org/protobuf/proto"
+)
+
+type Option func(r *Resolver)
+
+func WithHTTPClient(c *http.Client) Option {
+	return func(r *Resolver) {
+		r.httpClient = c
+	}
+}
+
+type Resolver struct {
+	httpClient *http.Client
+	cache      cache
+}
+
+func NewResolver(opts ...Option) *Resolver {
+	r := &Resolver{
+		httpClient: http.DefaultClient,
+		cache:      newCache(),
+	}
+	for _, opt := range opts {
+		opt(r)
+	}
+	return r
+}
+
+func (r *Resolver) Deref(ctx context.Context, ref *Ref) (io.ReadCloser, error) {
+	return r.deref(ctx, ref, true)
+}
+
+func (r *Resolver) deref(ctx context.Context, ref *Ref, cache bool) (io.ReadCloser, error) {
+	if data := r.cache.get(ref); data != nil {
+		return io.NopCloser(bytes.NewReader(data)), nil
+	}
+	cacheRC := func(rc io.ReadCloser) (io.ReadCloser, error) {
+		data, err := io.ReadAll(rc)
+		if err != nil {
+			return nil, err
+		}
+		r.cache.put(ref, data)
+		return io.NopCloser(bytes.NewReader(data)), nil
+	}
+	var rc io.ReadCloser
+	var err error
+	switch body := ref.Body.(type) {
+	case *Ref_Http:
+		rc, err = r.derefHTTP(ctx, body.Http)
+		if err != nil {
+			return nil, err
+		}
+	case *Ref_ContentHash:
+		rc, err = r.derefContentHash(ctx, body.ContentHash, false)
```

Comment:
Not sure I follow, the ref being resolved is cached if it is immutable. The caching will likely need to be improved because it currently caches a chunk for each transformation it undergoes, but that will just be a future improvement for the Go resolver. It isn't particularly important for the purpose of this PR, which is implementing the ReadFilesetCDR RPC. The Go resolver is just needed for testing at this point.